quantus/helpers/model/pytorch_model.py

```diff
@@ -54,27 +54,36 @@ def __init__(
             model_predict_kwargs=model_predict_kwargs,
         )
         self.device = device
-        self.cache = LRUCache(100)
 
-    @property
-    def _last_layer_is_softmax(self) -> bool:
+    @lru_cache(maxsize=None)
+    def _get_last_softmax_layer_index(self) -> Optional[int]:
         """
-        Checks if the last layer is an instance of torch.nn.Softmax.
+        Returns the index of the last module of torch.nn.Softmax type in the list of model children.
+        If no softmax module is found, returns None.
         """
+        modules = list(self.model.modules())
+        for i in range(-1, -len(modules), -1):
+            if isinstance(modules[i], torch.nn.Softmax):
+                return i
+        return None
+
         last_layer = list(self.model.children())[-1]
         return isinstance(last_layer, torch.nn.Softmax)
 
-    @cachedmethod(operator.attrgetter("cache"))
+    @lru_cache(maxsize=None)
     def _get_model_with_linear_top(self) -> torch.nn:
         """
-        In a case model has a softmax on top, and we want linear,
-        we have to rebuild the model and replace top with linear activation.
-        Cache the rebuilt model and reuse it during consecutive predict calls.
+        In a case model has a softmax module, the last torch.nn.Softmax module in the self.model.modules() list is
+        replaced with torch.nn.Identity().
         """
-        if not self._last_layer_is_softmax:
-            return self.model
+        linear_model = copy.deepcopy(self.model)
+
+        for attr_tpl in list(linear_model.named_modules())[::-1]:
```

Comment:
What does `attr_tpl` stand for? 
Please either use a more verbose self-descriptive variable name, or leave a comment on what does it mean.